model/map/src/main/java/org/keycloak/models/map/role/MapRoleProvider.java

```diff
@@ -0,0 +1,336 @@
+    }
+
+    @Override
+    public Stream<RoleModel> getRealmRolesStream(RealmModel realm, Integer first, Integer max) {
+        Stream<RoleModel> s = getRealmRolesStream(realm);
+        if (first != null && first >= 0) {
+            s = s.skip(first);
+        }
+        if (max != null && max >= 0) {
+            s = s.limit(max);
+        }
+        return s;
+    }
+
+    @Override
+    public Stream<RoleModel> getRealmRolesStream(RealmModel realm) {
+        return getNotRemovedUpdatedRolesStream()
+                .filter(entityRealmFilter(realm))
+                .sorted(COMPARE_BY_NAME)
+                .map(entityToAdapterFunc(realm));
+    }
+
+    @Override
+    public RoleModel addClientRole(ClientModel client, String id, String name) {
+        final UUID entityId = id == null ? UUID.randomUUID() : UUID.fromString(id);
+
+        LOG.tracef("addClientRole(%s, %s, %s)%s", client.getClientId(), id, name, getShortStackTrace());
+
+        MapRoleEntity entity = new MapRoleEntity(entityId, client.getId());
+        entity.setName(name);
+        entity.setClientRole(true);
+        entity.setClientId(client.getId());
+        entity.setRealmId(client.getRealm().getId());
+        if (tx.get(entity.getId(), roleStore::get) != null) {
+            throw new ModelDuplicateException("Role exists: " + id);
+        }
+        tx.putIfAbsent(entity.getId(), entity);
+        return entityToAdapterFunc(client.getRealm()).apply(entity);
+    }
+
+    @Override
+    public Stream<RoleModel> getClientRolesStream(ClientModel client, Integer first, Integer max) {
+        Stream<RoleModel> s = getClientRolesStream(client);
+        if (first != null && first >= 0) {
+            s = s.skip(first);
+        }
+        if (max != null && max >= 0) {
+            s = s.limit(max);
+        }
+        return s;
+    }
+
+    @Override
+    public Stream<RoleModel> getClientRolesStream(ClientModel client) {
+        return getNotRemovedUpdatedRolesStream()
+                .filter(entityRealmFilter(client.getRealm()))
+                .filter(entityClientFilter(client))
+                .sorted(COMPARE_BY_NAME)
+                .map(entityToAdapterFunc(client.getRealm()));
+    }
+    @Override
+    public boolean removeRole(RoleModel role) {
+        LOG.tracef("removeRole(%s(%s))%s", role.getName(), role.getId(), getShortStackTrace());
+
+        RealmModel realm = role.isClientRole() ? ((ClientModel)role.getContainer()).getRealm() : (RealmModel)role.getContainer();
+
+        session.users().preRemove(realm, role);
+
+        realm.removeDefaultRoles(role.getName());
+        
+        //remove role from realm-roles composites
+        try (Stream<RoleModel> baseStream = getRealmRolesStream(realm).filter(RoleModel::isComposite)) {
+            StreamUtils.leftInnerJoinStream(baseStream, RoleModel::getCompositesStream)
+              .filter(pair -> role.equals(pair.getV()))
+              .forEach(pair -> pair.getK().removeCompositeRole(role));
```

Comment:
When realm is being deleted I get `ConcurrentModificationException` here. It seems the problem is that while we iterate over underlying collection we also remove from it.

Reply:
Here a `.collect` is necessary before `forEach` since the stream of composite roles is still being read in the `RoleModel::getCompositesStream` above. Basically the same update as in https://github.com/keycloak/keycloak/pull/7436